qiskit/qpy/binary_io/value.py

```diff
@@ -82,6 +86,106 @@ def _write_parameter_expression(file_obj, obj):
         file_obj.write(value_data)
 
 
+class _ExprWriter(expr.ExprVisitor[None]):
+    __slots__ = ("file_obj", "clbit_indices")
+
+    def __init__(self, file_obj, clbit_indices):
+        self.file_obj = file_obj
+        self.clbit_indices = clbit_indices
+
+    def visit_generic(self, node, /):
+        raise exceptions.QpyError(f"unhandled Expr object '{node}'")
+
+    def visit_var(self, node, /):
+        self.file_obj.write(type_keys.Expression.VAR)
+        _write_expr_type(self.file_obj, node.type)
+        if isinstance(node.var, Clbit):
+            self.file_obj.write(type_keys.ExprVar.CLBIT)
+            self.file_obj.write(
+                struct.pack(
+                    formats.EXPR_VAR_CLBIT_PACK,
+                    *formats.EXPR_VAR_CLBIT(self.clbit_indices[node.var]),
+                )
+            )
+        elif isinstance(node.var, ClassicalRegister):
+            self.file_obj.write(type_keys.ExprVar.REGISTER)
+            self.file_obj.write(
+                struct.pack(
+                    formats.EXPR_VAR_REGISTER_PACK, *formats.EXPR_VAR_REGISTER(len(node.var.name))
+                )
+            )
+            self.file_obj.write(node.var.name.encode(common.ENCODE))
+        else:
+            raise exceptions.QpyError(f"unhandled Var object '{node.var}'")
+
+    def visit_value(self, node, /):
+        self.file_obj.write(type_keys.Expression.VALUE)
+        _write_expr_type(self.file_obj, node.type)
+        if node.value is True or node.value is False:
+            self.file_obj.write(type_keys.ExprValue.BOOL)
+            self.file_obj.write(
+                struct.pack(formats.EXPR_VALUE_BOOL_PACK, *formats.EXPR_VALUE_BOOL(node.value))
+            )
+        elif isinstance(node.value, int):
+            self.file_obj.write(type_keys.ExprValue.INT)
+            if node.value == 0:
+                num_bytes = 0
+                buffer = b""
+            else:
+                # This wastes a byte for `-(2 ** (8*n - 1))` for natural `n`, but they'll still
+                # decode fine so it's not worth another special case.  They'll encode to
+                # b"\xff\x80\x00\x00...", but we could encode them to b"\x80\x00\x00...".
+                num_bytes = (node.value.bit_length() // 8) + 1
+                buffer = node.value.to_bytes(num_bytes, "big", signed=True)
```

Comment:
Heh, I would have totally missed setting this explicitly. At least it looks like it defaults to big endian and we'd have matched the endianness for qpy.

I guess this is needed in case there is a >63 bit register being used in the expression so we can't use a fixed width integer encoding here.

Reply:
Yeah, that's the exact reason I put this conversion in.  `int.to_bytes` and `int.from_bytes` both require you to set the endianness, so I _had_ to remember lol.